### DISCUSSION TITLE
Yourls changes link

### DISCUSSION THREAD

**Discussion by mike14-2** [YOURLS/YOURLS#3741] [Q&A] [answered]
Hello,
I have the problem that I have a link which also contains capital letters. These must remain capitalized.
However, Yourls changes this link to be resolved.
Can anyone tell me what I need to do differently?
Thank you very much

**Comment by dgw** [comment]
Duplicate #2977?

**Reply by dgw** [reply]
No: OP commented extra details on #2977 describing a different issue that deserves its own thread.

I de-upvoted my own answer and will now hide it.

**Comment by dgw** [comment, marked as answer]
Based on additional details posted [in another thread](https://github.com/orgs/YOURLS/discussions/2977#discussioncomment-9122385), I believe the problem is that an application called DEVONthink misapplies the URI format in its own links as follows:

```
x-devonthink-item://930FBCF1-CB42-45B8-9092-904030947FBF
\________________/^^\__________________________________/
        |         |                  |
  custom scheme   |              path (UUID)
                  |
        authority prefix in URI
        that defines only a path
```

`//` immediately after the `scheme:` is reserved for an optional "authority" component—commonly, a hostname. In the next URI component, the "path", `//` can appear if a path segment is empty—but the first path segment may _not_ be empty if no authority is present in the URI. See more details in [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3) or in slightly plainer English [at Wikipedia](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax).

I think [this 5-year-old thread about DEVONthink links in Anki](https://discourse.devontechnologies.com/t/x-devonthink-item-links-in-anki/45366) means that Anki is parsing URIs correctly, and DEVONthink needs to use zero, one, or three slashes, but _not two_ for its links. If it is going to continue using URIs as they currently are, where the item URL is constructed as if the item UUID is a hostname, the item UUID should be treated as case-insensitive when the user clicks an item link.

Further detail of my investigation in the collapsible section below.

<details>
<summary>Extra analysis</summary>

Like [YOURLS' normalization function](https://github.com/YOURLS/YOURLS/blob/268aa5e0cc15f6e08fb85dad250295e92d318b45/includes/functions-formatting.php#L572-L651) written in PHP, the Python standard-library URI parser `urllib.parse.urlparse()` _also_ thinks this item UUID is a network location instead of a path:

```pycon
>>> from urllib.parse import urlparse
>>> res = urlparse('x-devonthink-item://930FBCF1-CB42-45B8-9092-904030947FBF')
>>> res
ParseResult(scheme='x-devonthink-item', netloc='930FBCF1-CB42-45B8-9092-904030947FBF', path='', params='', query='', fragment='')
```

As I suggested above, using any of the possible slash counts _other than_ the two that DEVONthink currently uses will work:

```pycon
>>> noslash = urlparse('x-devonthink-item:930FBCF1-CB42-45B8-9092-904030947FBF')
>>> noslash
ParseResult(scheme='x-devonthink-item', netloc='', path='930FBCF1-CB42-45B8-9092-904030947FBF', params='', query='', fragment='')
>>> oneslash = urlparse('x-devonthink-item:/930FBCF1-CB42-45B8-9092-904030947FBF')
>>> oneslash
ParseResult(scheme='x-devonthink-item', netloc='', path='/930FBCF1-CB42-45B8-9092-904030947FBF', params='', query='', fragment='')
>>> threeslash = urlparse('x-devonthink-item:///930FBCF1-CB42-45B8-9092-904030947FBF')
>>> threeslash
ParseResult(scheme='x-devonthink-item', netloc='', path='/930FBCF1-CB42-45B8-9092-904030947FBF', params='', query='', fragment='')
```

Zero slashes is the best choice if they do not want to handle the leading `/` of the resulting path, but three slashes is likely the most "normal-looking" to most users who have never paid attention to the exact number of slashes in the links they click every day.

</details>

**Reply by mike14-2** [reply]
Hello,
sorry for posting my answer incorrectly.
I have understood it, I think.
I now use “///” and it works.
Brilliant.

Thank you very much
